src/graph/unit_graph.cc

```diff
@@ -907,52 +908,68 @@ UnitGraph::UnitGraph(GraphPtr metagraph, CSRPtr in_csr, CSRPtr out_csr, COOPtr c
+  return !!std::atomic_load_explicit(&coo_, std::memory_order_acquire);
+}
+
 UnitGraph::CSRPtr UnitGraph::GetInCSR() const {
-  if (!in_csr_) {
-    if (out_csr_) {
-      const auto& newadj = aten::CSRTranspose(out_csr_->adj());
-      const_cast<UnitGraph*>(this)->in_csr_ = std::make_shared<CSR>(meta_graph(), newadj);
-    } else {
-      CHECK(coo_) << "None of CSR, COO exist";
-      const auto& adj = coo_->adj();
-      const auto& newadj = aten::COOToCSR(
-          aten::COOMatrix{adj.num_cols, adj.num_rows, adj.col, adj.row});
-      const_cast<UnitGraph*>(this)->in_csr_ = std::make_shared<CSR>(meta_graph(), newadj);
-    }
-  }
-  return in_csr_;
+  return LazyInit(&(const_cast<UnitGraph*>(this)->in_csr_), &lazy_mutex_, [this] () {
+      if (out_csr_) {
+        const auto& newadj = aten::CSRTranspose(out_csr_->adj());
+        return CSR(meta_graph(), newadj);
+      } else {
+        CHECK(coo_) << "None of CSR, COO exist";
+        const auto& adj = coo_->adj();
+        const auto& newadj = aten::COOToCSR(
+            aten::COOMatrix{adj.num_cols, adj.num_rows, adj.col, adj.row});
+        return CSR(meta_graph(), newadj);
+      }
+    });
 }
 
 /* !\brief Return out csr. If not exist, transpose the other one.*/
 UnitGraph::CSRPtr UnitGraph::GetOutCSR() const {
-  if (!out_csr_) {
-    if (in_csr_) {
-      const auto& newadj = aten::CSRTranspose(in_csr_->adj());
-      const_cast<UnitGraph*>(this)->out_csr_ = std::make_shared<CSR>(meta_graph(), newadj);
-    } else {
-      CHECK(coo_) << "None of CSR, COO exist";
-      const auto& newadj = aten::COOToCSR(coo_->adj());
-      const_cast<UnitGraph*>(this)->out_csr_ = std::make_shared<CSR>(meta_graph(), newadj);
-    }
-  }
-  return out_csr_;
+  return LazyInit(&(const_cast<UnitGraph*>(this)->out_csr_), &lazy_mutex_, [this] () {
+      if (in_csr_) {
+        const auto& newadj = aten::CSRTranspose(in_csr_->adj());
+        return CSR(meta_graph(), newadj);
+      } else {
+        CHECK(coo_) << "None of CSR, COO exist";
+        const auto& newadj = aten::COOToCSR(coo_->adj());
+        return CSR(meta_graph(), newadj);
+      }
+    });
 }
 
 /* !\brief Return coo. If not exist, create from csr.*/
 UnitGraph::COOPtr UnitGraph::GetCOO() const {
-  if (!coo_) {
-    if (in_csr_) {
-      const auto& newadj = aten::CSRToCOO(in_csr_->adj(), true);
-      const_cast<UnitGraph*>(this)->coo_ = std::make_shared<COO>(
-          meta_graph(),
-          aten::COOMatrix{newadj.num_cols, newadj.num_rows, newadj.col, newadj.row});
-    } else {
-      CHECK(out_csr_) << "Both CSR are missing.";
-      const auto& newadj = aten::CSRToCOO(out_csr_->adj(), true);
-      const_cast<UnitGraph*>(this)->coo_ = std::make_shared<COO>(meta_graph(), newadj);
-    }
-  }
-  return coo_;
+  return LazyInit(&(const_cast<UnitGraph*>(this)->coo_), &lazy_mutex_, [this] () {
+      if (in_csr_) {
+        const auto& newadj = aten::CSRToCOO(in_csr_->adj(), true);
+        return COO(
```

Comment:
Will this cause extra data copy?

Reply:
Which line are you referring to?

Reply:
L964

Reply:
It shouldn't?  Because in `LazyInit()` this is where the lambda function is called:
```c++
result.reset(new T(func()))
```
So the COO object would be moved to the shared pointer and there shouldn't be any copy.

Correct me if I'm wrong though, as I'm sometimes also confused by when C++ copies and moves an object...